packages/transaction-controller/src/TransactionController.ts

```diff
@@ -2676,4 +2682,25 @@ export class TransactionController extends BaseControllerV1<
       log('Error while updating post transaction balance', error);
     }
   }
+
+  private async publishTransactionWithValidation(
+    rawTx: string,
+    transactionMeta: TransactionMeta,
+  ): Promise<string> {
+    let hash;
+    try {
+      hash = await this.publishTransaction(rawTx);
+      // TODO: Replace `any` with type
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    } catch (error: any) {
+      if (error?.code === errorCodes.rpc.invalidInput) {
+        await this.pendingTransactionTracker.isTransactionConfirmed(
+          transactionMeta,
+        );
+        throw new Error('Transaction already confirmed.');
+      }
+      throw new Error('Failed to publish transaction.');
```

Comment:
For the sake of clarity, if it wasn't this already confirmed error, shall we throw the original error again so we don't impact more flows than intended?

Reply:
Good point, fixed in https://github.com/MetaMask/core/pull/3800/commits/84cf788287b53772c57b770eee99bc5818da3598